File: 2020/scala/src/main/scala/Advent01.scala
import AdventApp.ErrorMessage

object Advent01 extends SingleLineAdventApp[Int, Long]:
  def fileName: String = "01.txt"

  def solve(testCases: List[Int], n: Int): Long =
    testCases
      .combinations(n)
      .find { _.sum == 2020 }
      .getOrElse(sys.error("Failed"))
      .product

  def solution1(testCases: List[Int]): Long = solve(testCases, 2)
  def solution2(testCases: List[Int]): Long = solve(testCases, 3)
  
  def parseLine(line: String): Either[ErrorMessage, Int] = 
    line.toIntOption.toRight(ErrorMessage("Failed to parse"))

File: 2020/haskell/Advent01.hs
module Advent01 where

combinations :: Int -> [a] -> [[a]]
combinations 0 _ = [[]]
combinations _ [] = []
combinations n (x : xs) = map (x :) (combinations (n - 1) xs) ++ combinations n xs

toInt :: String -> Int
toInt x = read x :: Int

solve :: Int -> [Int] -> Int
solve n list = product $ head $ filter (\x -> sum x == 2020) (combinations n list)

main :: IO ()
main = do
  raw <- readFile "01.txt"
  let input = toInt <$> lines raw
  print $ solve 2 input
  print $ solve 3 input
